# 条款 7：预先准备好内存不够的情况

## 0. 总则

写代码时去考虑处理内存分配失败是一个很有礼貌而道德的事情，因为大多情况向下，如果内存都分配失败了，那么实际上程序没有多少继续运行的意义了。经本人测试，当不断地向系统申请分配内存时，如果忽略异常，最终系统会蓝屏。如果我们希望做一个谦谦君子，又不希望在每一个`new`的外面包一层`try`，最简单暴力快捷的方法是在程序启动后利用`set_new_handler`设置自定义异常函数；如果需要细化每个对象，则建议使用`NewHandleSupport`模版基类。

## 1. 自定义`new`的出错函数

> 当内存分配请求不能满足时，调用你预先指定的一个出错处理函数。这个方法基于一个常规，即当 `operator new`不能满足请求时，会在抛出异常之前调用客户指定的一个出错处理函数——一般称为`new-handler`函数。

在C++中，指定出错处理函数时要用到`set_new_handler`函数，它在头文件`<new>`里大致是象下面这样定义的：

```cpp
typedef void (*new_handler)();
new_handler set_new_handler(new_handler p) throw();
```

经测试，该函数是在`operator new`失败时，但在抛出异常之前调用的，且如果设置了自定义的函数，就不会再抛出异常了。

```cpp
#include <iostream>
#include <new>

bool    bRun = true;
char*   pReserve = NULL;
void MyNewHandle()
{
    std::cout << "Hello，World." << std::endl;
    bRun = false;
    
    // 释放预留空间
    delete[] pReserve;
    pReserve = NULL;
}
int main()
{
    std::set_new_handler(MyNewHandle);
    try
    {
        pReserve = new char[1024 * 1024 * 100];
        while (bRun)
        {
            new long(1);
        }
        std::cout << "没有异常异常，正常结束" << std::endl;
    }
    catch (std::bad_alloc& e)
    {
        std::cout << "bad_alloc异常:" << e.what() << std::endl;
    }

    return 0;
}
```

上述例子输出结果为:

> Hello，World.  
> 没有异常异常，正常结束

并没有扑捉到异常，且上述示例中预分配的内存是为了在内存分配失败时，释放该内存使得下次`new`可以有更多的可选空间。此外，如果没有该预分配的空间并在失败时释放，那么此次调用失败的new并不会结束，实际上会在调用函数`MyNewHandle`后继续执行`new`，而不会执行下一个`while`判断，为了验证该情况可以尝试将释放预留空间的代码注释掉再运行看看，注意打断点方法。

自定义`new`的出错函数一般要实现的功能:

* 产生更多的内存，如释放预分配的内存。
* 安装另一个错误处理函数。
* 卸除 new-handler，置空后会继续抛出异常。
* 抛出std::bad_alloc。
* 没有返回，程序报错或退出。

## 2. 自定义`new`异常处理的模版基类`NewHandleSupport`

```cpp
#include <iostream>
#include <new>

#define SHOW_CUR_TAG std::cout << "TAG INFO:" << __FUNCTION__ << "," << __LINE__ << std::endl;

template<class T>
class NewHandleSupport
{
public:
    // 设置本类的异常处理函数，要求每个类对象使用前至少调用一次
    static new_handler set_new_handler(new_handler p);

    // 重载本类的内存创建函数，保证自定义new的流程
    static void* operator new(size_t nSize);

private:
    // 保存本类的new失败异常处理函数
    static new_handler m_pCurNewHandle;
};

template<class T>
new_handler NewHandleSupport<T>::m_pCurNewHandle = NULL;

template<class T>
new_handler NewHandleSupport<T>::set_new_handler(new_handler p)
{
    new_handler pOldHanldel = m_pCurNewHandle;
    m_pCurNewHandle = p;
    return pOldHanldel;
}

template<class T>
void* NewHandleSupport<T>::operator new(size_t nSize)
{
    new_handler pOldHandle = std::set_new_handler(m_pCurNewHandle);
    void* pDat = NULL;

    try
    {
        // 标准new函数
        pDat = ::operator new(nSize);
    }
    catch (std::bad_alloc&)
    {
        // 如果m_pCurNewHandle指定的异常函数不为空，
        // 且异常能够腾出更多的空间或者程序异常退出，那么本处异常不会执行
        SHOW_CUR_TAG;
        std::set_new_handler(pOldHandle);
        throw;
    }
    std::set_new_handler(pOldHandle);
    return pDat;
}

bool    g_bRunA = true;
char*   g_pReserve = NULL;
void MyHandleA()
{
    std::cout << "A-Game Over." << std::endl;

    // 释放预留空间，腾出空间给new使用
    delete[] g_pReserve;
    g_pReserve = NULL;
    g_bRunA = false;
}

class A : public NewHandleSupport<A>
{
public:
    char a[1024];
    int  b[1024];
};

class B : public NewHandleSupport<B>
{
public:
    char a[1024];
    int  b[1024];
};

int main()
{
    g_pReserve = new char[1024 * 1024 * 512];
    // 要求每个类自定义初始化异常处理函数，否则会抛出异常
    // 此处A已初始化，不会抛出异常
    A::set_new_handler(MyHandleA);
    while (g_bRunA)
    {
        new A();
    }

    // 此处B未初始化异常函数，会抛出异常
    while (true)
    {
        new B();
    }
    return 0;
}
```

输出结果为:

> A-Game Over.  
> TAG INFO:NewHandleSupport<class B>::operator new,47